sdk/include/reactos/DX_Shell.h

```diff
@@ -0,0 +1,221 @@
+{
+    INT nLength = strPath.GetLength();
+    if (nLength > 0 && strPath[nLength - 1] == L'\\')
+        return;
+    strPath += L'\\';
+}
+
+inline VOID DX_PathRemoveBackslashW(CStringW& strPath)
+{
+    INT nLength = strPath.GetLength();
+    if (nLength > 0 && strPath[nLength - 1] == L'\\')
+        strPath = strPath.Left(nLength - 1);
+}
+
+inline VOID DX_PathAppend(CStringW& strPath, LPCWSTR pszAppend)
+{
+    DX_PathAddBackslashW(strPath);
+    strPath += pszAppend;
+}
+
+inline VOID DX_PathRemoveFileSpecW(CStringW& strPath)
+{
+    LPWSTR pszPath = strPath.GetBuffer();
+    PathRemoveFileSpecW(pszPath);
+    strPath.ReleaseBuffer();
+}
+
+inline VOID DX_PathAddExtensionW(CStringW& strPath, LPCWSTR pszDotExt)
+{
+    strPath += pszDotExt;
+}
+
+inline VOID DX_PathRemoveExtensionW(CStringW& strPath)
+{
+    LPWSTR pszPath = strPath.GetBuffer();
+    PathRemoveExtensionW(pszPath);
+    strPath.ReleaseBuffer();
+}
+
+inline BOOL DX_PathStripToRootW(CStringW& strPath)
+{
+    BOOL ret = TRUE;
+
+    LPWSTR lpszPath = strPath.GetBuffer();
+    while (!PathIsRootW(lpszPath))
+    {
+        if (!PathRemoveFileSpecW(lpszPath))
+        {
+            ret = FALSE;
+            break;
+        }
+    }
+    strPath.ReleaseBuffer();
+
+    return ret;
+}
+
+inline BOOL DX_PathCanonicalizeW(CStringW& strDest, LPCWSTR lpszPath)
+{
+    strDest = lpszPath;
+    LPWSTR pszBuffer = strDest.GetBuffer();
+    BOOL ret = PathCanonicalizeW(pszBuffer, lpszPath);
+    strDest.ReleaseBuffer();
+    return ret;
+}
+
+inline VOID DX_PathCombineW(CStringW& strDest, LPCWSTR lpszDir, LPCWSTR lpszFile)
+{
+    if (!lpszDir && !lpszFile)
+    {
+        strDest.Empty();
+        return;
+    }
+
+    BOOL bUseBoth = FALSE, bStrip = FALSE;
+
+    CStringW strTemp;
+    if ((!lpszFile || !*lpszFile) && lpszDir)
+    {
+        strTemp = lpszDir; // Use dir only
+    }
+    else if (!lpszDir || !*lpszDir || !PathIsRelativeW(lpszFile))
+    {
+        if (!lpszDir || !*lpszDir || *lpszFile != L'\\' || DX_PathIsUNCW(lpszFile))
+            strTemp = lpszFile; // Use file only
+        else
+            bUseBoth = bStrip = TRUE;
+    }
+    else
+    {
+        bUseBoth = TRUE;
+    }
+
+    if (bUseBoth)
+    {
+        strTemp = lpszDir;
+        if (bStrip)
+        {
+            DX_PathStripToRootW(strTemp);
+            lpszFile++; // Skip '\'
+        }
+
+        DX_PathAddBackslashW(strTemp);
+        strTemp += lpszFile;
+    }
+
+    DX_PathCanonicalizeW(strDest, strTemp);
+}
+
+inline VOID DX_PathQuoteSpacesW(CStringW& strPath)
+{
+    CStringW strTemp = L"\"";
+    strTemp += strPath;
+    strTemp += L'\"';
+    strPath = strTemp;
+}
+
+inline VOID DX_PathUnquoteSpacesW(CStringW& strPath)
+{
+    if (strPath[0] == L'\"' && strPath[strPath.GetLength() - 1] == L'\"')
+        strPath = strPath.Mid(1, strPath.GetLength() - 2);
+}
+
+inline HRESULT DX_SHGetPathFromIDListW(LPCITEMIDLIST pidl, LPWSTR pszPath, UINT cchPath)
```

Comment:
Is this basically some sort of copy-paste of the existing SHGetPathFromIDListW function? (and same question for the two other functions below)
If so I'm not sure it is a good practice to proceed this way.

Reply:
Yes except it has a buffer length parameter and the return value is an `HRESULT`.

Reply:
See https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetpathfromidlistex

Reply:
That is Vista+ function.

Reply:
Maybe but if you use it only internally (only within shell32 itself), there shouldn't be any problem.

Reply:
I added `SHGetPathFromIDListDxW` as an extension of `SHGetPathFromIDListW`.